internal/graph/check.go

```diff
@@ -193,26 +188,29 @@ func (cc *ConcurrentChecker) checkDirect(ctx context.Context, crc currentRequest
 	defer it.Close()
 
 	// Find the subjects over which to dispatch.
-	foundResourceIds := []string{}
+	foundResources := NewMembershipSet()
 	subjectsToDispatch := tuple.NewONRByTypeSet()
-	resourceIDsBySubjectID := util.NewMultiMap[string, string]()
+	relationshipsBySubjectONR := util.NewMultiMap[string, *core.RelationTuple]()
 
 	for tpl := it.Next(); tpl != nil; tpl = it.Next() {
 		if it.Err() != nil {
 			return checkResultError(NewCheckFailureErr(it.Err()), emptyMetadata)
 		}
 
+		// If the subject of the relationship matches the target subject, then we've found
+		// a result.
 		if onrEqualOrWildcard(tpl.Subject, crc.parentReq.Subject) {
-			foundResourceIds = append(foundResourceIds, tpl.ResourceAndRelation.ObjectId)
-			if crc.resultsSetting == v1.DispatchCheckRequest_ALLOW_SINGLE_RESULT {
-				return checkResultsForResourceIds(foundResourceIds, emptyMetadata)
+			foundResources.AddDirectMember(tpl.ResourceAndRelation.ObjectId, tpl.Caveat)
+			if crc.resultsSetting == v1.DispatchCheckRequest_ALLOW_SINGLE_RESULT && foundResources.HasDeterminedMember() {
+				return checkResultsForMembership(foundResources, emptyMetadata)
```

Comment:
It may be faster to resolve the caveat here at the leaf if `HasDeterminedMember == false` than dispatching.

One possible optimization would be to hedge the dispatch against the caveat evaluation:
- concurrently start evaluating all caveated members
- concurrently send dispatch request
- if any of the relationships with caveats resolve to `MEMBER` cancel the dispatched request.

I don't think we need to optimize this right now but we should keep a list of potential optimizations to start working on once we have the first end to end functional example working.

Reply:
We can't resolve the caveat here without passing the full context into the DispatchCheckRequest, which means that the cache key would need to include the full context, thus making cache reuse much less likely here.

We also cannot return immediately because if we do not have a determined result, we cannot know for certain that this will short circuit

Reply:
That is certainly an important consideration. While early to say, "I have the sneaking suspicion" that operations like lookup* with a large corpus of caveated tuples would become CPU bound due to the execution of potential complex expressions that has to be carried out for the ingress SpiceDB node. That single node becomes the bottleneck for the streaming API. In contrast, evaluating in the owning node of the ring means we are can continue to scale by trowing more compute at the problem. 

So it may be a worthwhile investment to include the context in the dispatch request and it in the cache, at the expense of higher memory requirements and potential less hit rate if it avoids costly dispatches. And I guess we only need to do this if caveated tuples where found?